File: zad_4.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>

using namespace std;

// Метод Борда
string borda_method(const vector<vector<string>>& votes, const vector<string>& candidates) {
    unordered_map<string, int> score;
    int n = candidates.size();

    for (const auto& vote : votes) {
        for (int i = 0; i < n; ++i) {
            score[vote[i]] += n - i - 1;
        }
    }

    return max_element(score.begin(), score.end(),
        [](const auto& a, const auto& b) { return a.second < b.second; })->first;
}

// Метод Кондорсе
string condorcet_method(const vector<vector<string>>& votes, const vector<string>& candidates) {
    int n = candidates.size();
    unordered_map<string, unordered_map<string, int>> wins;

    for (const auto& vote : votes) {
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                wins[vote[i]][vote[j]]++;
            }
        }
    }

    for (const auto& cand1 : candidates) {
        bool winner = true;
        for (const auto& cand2 : candidates) {
            if (cand1 == cand2) continue;
            if (wins[cand1][cand2] <= wins[cand2][cand1]) {
                winner = false;
                break;
            }
        }
        if (winner) return cand1;
    }
    return "Нет победителя по Кондорсе";
}

int main() {
    int n, k;
    cout << "Введите количество кандидатов и избирателей: ";
    cin >> n >> k;

    vector<string> candidates(n);
    cout << "Введите имена кандидатов:\n";
    for (int i = 0; i < n; ++i) {
        cin >> candidates[i];
    }

    vector<vector<string>> votes(k, vector<string>(n));
    cout << "Введите предпочтения избирателей (по " << n << " кандидатов в строке):\n";
    for (int i = 0; i < k; ++i) {
        for (int j = 0; j < n; ++j) {
            cin >> votes[i][j];
        }
    }

    string borda_winner = borda_method(votes, candidates);
    string condorcet_winner = condorcet_method(votes, candidates);

    cout << "\nПобедитель по методу Борда: " << borda_winner << "\n";
    cout << "Победитель по методу Кондорсе: " << condorcet_winner << "\n";

    if (borda_winner != condorcet_winner && condorcet_winner != "Нет победителя по Кондорсе") {
        cout << "Разные алгоритмы дали разные результаты.\n";
    }

    return 0;
}

File: zad_4.go
package main

import (
 "bufio"
 "fmt"
 "os"
 "strings"
)

func readLine(scanner *bufio.Scanner) string {
 scanner.Scan()
 return scanner.Text()
}

// Метод Борда
func borda(votes [][]string, candidates []string) string {
 score := make(map[string]int)
 n := len(candidates)

 for _, vote := range votes {
  for i, cand := range vote {
   score[cand] += n - i - 1
  }
 }

 maxScore := -1
 var winner string
 for cand, s := range score {
  if s > maxScore {
   maxScore = s
   winner = cand
  }
 }
 return winner
}

// Метод Кондорсе
func condorcet(votes [][]string, candidates []string) string {
 n := len(candidates)
 winCount := make(map[string]map[string]int)

 // Инициализация
 for _, a := range candidates {
  winCount[a] = make(map[string]int)
  for _, b := range candidates {
   winCount[a][b] = 0
  }
 }

 for _, vote := range votes {
  for i := 0; i < n; i++ {
   for j := i + 1; j < n; j++ {
    winCount[vote[i]][vote[j]]++
   }
  }
 }

 for _, cand1 := range candidates {
  winner := true
  for _, cand2 := range candidates {
   if cand1 == cand2 {
    continue
   }
   if winCount[cand1][cand2] <= winCount[cand2][cand1] {
    winner = false
    break
   }
  }
  if winner {
   return cand1
  }
 }
 return "Нет победителя по Кондорсе"
}

func main() {
 scanner := bufio.NewScanner(os.Stdin)

 fmt.Print("Введите количество кандидатов и избирателей: ")
 var n, k int
 fmt.Scanf("%d %d\n", &n, &k)

 fmt.Println("Введите имена кандидатов:")
 scanner.Scan()
 candidates := strings.Fields(scanner.Text())

 votes := make([][]string, k)
 fmt.Println("Введите предпочтения избирателей (по", n, "кандидатов в строке):")
 for i := 0; i < k; i++ {
  scanner.Scan()
  votes[i] = strings.Fields(scanner.Text())
 }

 bordaWinner := borda(votes, candidates)
 condorcetWinner := condorcet(votes, candidates)

 fmt.Println("\nПобедитель по методу Борда:", bordaWinner)
 fmt.Println("Победитель по методу Кондорсе:", condorcetWinner)

 if bordaWinner != condorcetWinner && condorcetWinner != "Нет победителя по Кондорсе" {
  fmt.Println("Разные алгоритмы дали разные результаты.")
 }
}
